create persian.sbl

saeiddrv opened the pull request.

Commit:
create persian.sbl

saeiddrv commented:
add Persian stemming algorithm!

ojwb commented on python/modules.txt:
```diff
@@ -0,0 +1,62 @@
+lithuanian      UTF_8                   lithuanian,lt,lit
+nepali          UTF_8                   nepali,ne,nep
+norwegian       UTF_8,ISO_8859_1        norwegian,no,nor
+persian         UTF_8                   persian,fa,fas,pers
```
This line looks right, but you should add it to the existing `libstemmer/modules.txt` rather than creating a new version of the file in a different subdirectory.

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,117 @@
+define Normalize as (
+    do repeat (
+        [substring] among (
+            '{U+06A9}' '{U+0643}' ( <- '{kaf}' ) // Normalize Kaf
```
From above:

```
stringdef kaf     '{U+06A9}'
```

So as part of this you seem to be replacing kaf with itself, which seems pointless.

Commit:
fix errors

Commit:
add exception section

Commit:
fix exception section

Commit:
add Suffix_Normalize

Commit:
fix modules.txt

saeiddrv commented:
I have fixed the issues.

ojwb commented on algorithms/persian.sbl:
```diff
@@ -115,34 +112,34 @@ define Suffix_Adjective as (
+        '{sin}{lam}{alef}{mim}' ()
+        '{alef}{sin}{lam}{alef}{mim}' ()
+        '{ye}{ain}{nun}{ye}' ()
+        '{aa}{lam}{be}{vav}{mim}' ()
```
There's no need for all these explicit empty actions - no action is the default.

ojwb commented on algorithms/persian.sbl:
```diff
@@ -55,11 +54,11 @@ externals ( stem )
+            '{U+0629}' ( <- '{he}'  )             // Normalize Heh
             '{U+0623}' '{U+0625}' ( <- '{alef}' ) // Normalize Alef
-            '{U+0624}' ( <- '{vav}' )  // Normalize Waw and Waw with Hamza above
+            '{U+0624}' ( <- '{vav}' )             // Normalize Waw and Waw with Hamza above
```
As I noted before, it'd be better to use stringdef names for everything.

Then it's clearer to the reader what the code is doing.  Consider the line:

```
            '{U+0623}' '{U+0625}' ( <- '{alef}' ) // Normalize Alef
```

Currently I need to e.g. look up U+0623 to find it's "ARABIC LETTER ALEF WITH HAMZA ABOVE", then look up U+0625 to find it's "ARABIC LETTER ALEF WITH HAMZA BELOW" before I can fully understand this single line.  If it instead read something like this then it would be clear without additional research:

```
            '{alef_with_hamza_above}' '{alef_with_hamza_below}' ( <- '{alef}' ) // Normalize Alef
```

ojwb commented on algorithms/persian.sbl:
```diff
@@ -115,34 +112,34 @@ define Suffix_Adjective as (
 
 define stem as (
-    Exception
+    ( Exception )
```
This means the same with or without the parentheses (just noting as the commit that changed this was "fix exception section"...)

Commit:
define arabic characters

ojwb commented on libstemmer/modules.txt:
```diff
@@ -29,6 +29,7 @@ italian         UTF_8,ISO_8859_1        italian,it,ita
 lithuanian      UTF_8                   lithuanian,lt,lit
 nepali          UTF_8                   nepali,ne,nep
 norwegian       UTF_8,ISO_8859_1        norwegian,no,nor
+persian         UTF_8                   persian,fa,fas,pers
```
I think that should be `per` not `pers` at the end (it should be a three letter code and https://en.wikipedia.org/wiki/List_of_ISO_639_language_codes lists `per` as the ISO 639-2/B code for Persian).

ojwb commented:
> We also need test data and a description for a new algorithm - there's a detailed description of the requirements in `CONTRIBUTING.rst`.

@saeiddrv You've since opened https://github.com/snowballstem/snowball-data/pull/25 to add test data - thanks for doing that.

We really do still need a description of the new algorithm though, which should be a pull request against the https://github.com/snowballstem/snowball-website repo.

We've found that algorithms without good documentation are much harder to maintain - for example, if somebody reports that a particular stemmer doesn't remove a prefix or suffix it's a usually lot of work to try to determine if that's an oversight or something that was considered but rejected for a good reason (most languages seem to have a prefix or suffix that can't be safely removed without causes worse problems in words that start or end with the same letters but where they aren't the intended prefix/suffix), but if we have a document that says "suffixes A, B and C are not removed because of X and Y" then we can just point the bug reporter to that.

If the implementation is from an algorithm that's described in a published paper we may be able to link to that to provide most of the information, provided the paper does contain a sufficiently detailed description of the algorithm and discussion of the design decisions that lead to it.

I can help with proof-reading and the like, but I didn't develop the algorithm and I don't know the Persian language so it isn't a document I could write the actual content of.

https://snowballstem.org/algorithms/indonesian/stemmer.html is a reasonable example showing the sort of thing that's useful.  There are macros to help render the list of references, the table of example stems, and the snowball code - see `algorithms/indonesian/stemmer.tt` in the repo.

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,181 @@
+            '{ar_vav_with_hamza_above}' ( <- '{vav}' )
+
+
+            '{gaf}{alef}{nun}' ( <- '{he}'  )     // gaan -> he
```
I suspect you want to add `'' (next)` here, assuming it's meant to replace any instance of one of these characters in the word.

As currently written it will stop transliterating characters when it gets to a character which there isn't a transliteration for.

saeiddrv committed:
fix language code

saeiddrv committed:
fix using next statement

saeiddrv committed:
add more exceptions

saeiddrv commented:
@ojwb Thanks for your review, guidance, and patience! I will fix those issues and work on making it more accurate in my next pushes.

saeiddrv committed:
improve steps

saeiddrv committed:
fix Normalize_Nouns

saeiddrv committed:
fix Exceptions

saeiddrv committed:
update header comment

ojwb commented on algorithms/persian.sbl:
```diff
@@ -65,18 +65,17 @@ externals ( stem )
 
 
-            '{gaf}{alef}{nun}' ( <- '{he}'  )     // gaan -> he
+            '{gaf}{alef}{nun}' ( <- '{he}' next )     // gaan -> he
```
This doesn't seem to address the point I raised.

On the assumption that this routine is meant to loop over the whole string making these changes to any occurrences it finds then, I suggested adding a new `among` case:

```
            '' (next)
```

The longest matching string in the `among` is chosen and an empty string is always a match, so this acts kind of like a `default` case in a C switch.  Otherwise `repeat` will only loop until the `among` doesn't match, so it will only change things in the initial span of the word which is made up of strings in the `among`.

Instead you've added `next` after every action - the `among` matching already advances the cursor over the matched string (and `<- 'x'`/`delete` in the action will adjust the cursor to still be on the character after the matched string), so adding `next` like you have means that this routine won't perform two adjacent replacements.  If that's not possible for any Persian words then that's OK (and doing `next` after a replacement is effectively an optimisation), but you still need to advance in the case where the `among` does not match.

ojwb commented:
Hmm, it seems you've made updates to fix my comments, but github still shows my comment last so I'd missed this until I noticed the "Outdated" marker on my comment.

I'm currently focusing on getting a new release out (it's very nearly 3.5 years since the last one which is too long and there's a huge pile of unreleased useful changes), and my plan is to then dealing with the pending new algorithms (including this one) and making another release with those.

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,287 @@
+stringdef ar_num9 '{U+0669}'
+
+// Other Characters
+stringdef half_space  '{U+200C}'
```
This seems wrong as `UnicodeData.txt` says `U+200C` is:

```
200C;ZERO WIDTH NON-JOINER;Cf;0;BN;;;;;N;;;;;
```

Not sure if the `half_space` name is wrong or the hex code is wrong though.

Generally I'd expect the tokenisation stage which calls the stemmer to split on any space characters, so we shouldn't see anything that's `Zs` in `UnicodeData.txt`.  I don't see anything named `HALF SPACE` but e.g. I wouldn't expect the stemmer to see:

```
2009;THIN SPACE;Zs;0;WS;<compat> 0020;;;;N;;;;;
```

So if the hex code is wrong and should have been `U+2009` or similar, I'd say this is not something we need to check for.

dscorbett replied:
U+200C ZWNJ is called “<span lang="fa">نیم فاصله</span>” (‘half space’) in Persian. It’s a word-internal morpheme separator.

ojwb replied:
I see (that seems a misleading name as it's zero width not half width and isn't a space).

I can't see much evidence of U+200C also being known as _half space_ in English (there are some web search results but they all seem to be referring to the literal English translation of the Persian name).  So if we're going to refer to it as _half space_ here, I think we need to explain why in a comment.

saeiddrv committed:
Add explanatory comment for U200C half-space usage in Persian

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,291 @@
+            '{mim}{nun}{dal}' (delete)    // mand
+            '{nun}{alef}{kaf}' (delete)   // naak
+            '{gaf}{ye}' (delete)          // gii
+            '{alef}{nun}{heh}' (delete)   // aaneh
```
If an input word is exactly one of these suffixes, then the output stem is an empty string which seems better avoided even if none of these are actually valid words.

Checking `output.txt` from the snowball-data PR, there appear to be 155 entries in `voc.txt` which give an empty stem, at least some of which seem to be real words - e.g. _بهانه‌هایی_ which google translate says means "excuses".  We definitely don't want to stem real words to an empty stem.

If we should require leaving at least one character before the removed suffix, that can be easily (and efficiently) achieved by adding a check that the cursor isn't "`atlimit`" after the `substring` above like so:

```
[substring] not atlimit among (
```

If we should require leaving multiple characters then `$(len > X)` checks as used elsewhere are OK, if a bit inelegant (and less efficient for UTF-8 where the string length needs computing each time).

ojwb replied:
https://en.wikipedia.org/wiki/Persian_alphabet says there are "32 letters [in] the modern Persian alphabet" - that means there are only 32 possible different single letter stems, so producing a single letter stem for a longer input (particular for real words) is probably also a design error because it's very likely to result in unwanted conflation.

saeiddrv committed:
update persian algorithm based on the HPS

ojwb committed:
Merge branch 'master' into add-persian-stemmer

ojwb committed:
Simplify Strip_Prefix_At_Start

We don't need to mess around with getmark and setmark as Snowball's
default handling of the cursor does what we want here.

Replace use of an explicit `changed` flag with just using the signal
from Strip_Prefix_At_Start to indicate if it made a change.

ojwb committed:
Eliminate freeze_pass boolean

We can use the signal from Stem_Noun instead.

ojwb committed:
Eliminate changed boolean

We can just use the signals from the routines.

ojwb committed:
Eliminate temp_mark completely

ojwb committed:
Fix latent bugs in Protect_Lexical_AN

ojwb committed:
Fix numbering of steps

ojwb committed:
Remove unused setting of slice

ojwb committed:
Change to/from backwardmode less

There's a small overhead to each such change.

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,411 @@
+    [substring] among (
+      '{ye}{alef}{nun}'     ($(len > 5) delete set changed)                 // -YAN
+      '{gaf}{alef}{nun}'    ($(len > 5) delete set changed)                 // -GAN
+      '{heh}{alef}{ye}{ye}' ($(len > 6) delete set changed)                 // -HAYI
```
It looks to me like this rule can never match since any word which could match here ends `{ye}{ye}` and has `$(len > 4)` so will have matched the rule in `Noun_Other` on line 269 below.

If I comment out line 250 the output of the stemmer isn't affected.

ojwb committed:
FIXME comment redundant rule

saeiddrv committed:
Improve Persian stemming algorithm

ZWNJ-gated prefix stripping:
- Prefixes (می‌/نمی‌/نا‌/بی‌) now require a ZWNJ between prefix and stem,
  preventing false matches on بیرون، بیشتر، میلادی، میان etc.
- Add Delete_ZWNJ routine to clean up remaining ZWNJs after prefix detection
- Reset saw_present_prefix at start of stem to prevent flag leaking
  between words when the stemmer is called repeatedly on the same env

Lexical -AN guard extended:
- Add atlimit to all AN_LexHash entries for exact whole-word matching
  (fixes false protection of فرزندان due to زندان suffix match)
- Add {vav}{alef}{nun} (وان) protection pattern: عنوان، دیوان، جوان etc.
- Expand AN_LexHash from 9 to 41 entries: آسمان، سازمان، پایان، جریان،
  میزان، زندان، قهرمان، اصفهان، یونان، لبنان، کهکشان، آتشفشان، پریشان،
  همزمان، ساختمان، سلیمان and others
- Extend not-protect_lex_an guard to -YAN, -GAN, -MAN, -TAN, -SHAN and
  to the infinitive -AN rule in Verb_Tense_Mood_Markers

Remove V_Hash:
- The single entry {alef}{sin}{te} → {heh}{sin}{te} was wrongly replacing
  است (is/was) with هست (there exists); removed entirely

Conservative -ه stripping:
- Replace broad {heh} rule with test '{dal}' or test '{te}' check:
  only strip final -ه when preceded by د or ت (past-tense consonants),
  preserving nouns/adjectives ending in ه (گروه، خانه، الله etc.)
- Add {nun}{dal}{heh} (-نده) rule to strip agent nouns to present root:
  نویسنده → نویس، گیرنده → گیر

Remove over-aggressive suffix rules:
- Remove single-char possessives -م/-ت/-ش from Noun_Possessive
- Remove -م/-د from Participle_Clitic_Tails
- Remove -DAN ({dal}{alef}{nun}) from Noun_Other

Fix Verb_Person_Endings dead-code warning:
- Merge two [substring] among blocks into one, placing the verb-cue
  guard (saw_present_prefix or saw_inf_or_part) inside each generic
  entry action; eliminates Snowball compiler warning and 3 C warnings

Enable Stem_Verb:
- Remove FIXME comment and uncomment 'or Stem_Verb' in main stem loop

Fix comment inaccuracies:
- Adjective_Relative: remove incorrect mention of 'final -H cleanups'
- AN_LexHash: correct comment from 'exact matches' to 'suffix list'

ojwb committed:
Factor out repeated among action

ojwb committed:
Remove redudant whole word -an exceptions

All of these are handled by a suffix checked for in Protect_Lexical_AN.

For example `iran` is caught by `…ran`.

ojwb committed:
Optimise Delete_ZWNJ

Use the same approach as `postlude` in english.sbl.

ojwb committed:
Hoist `not protect_lex_an` checks

Checking a flag is very cheap, so do it first.  If it's true we know
the stem currently ends -an so it can't match suffixes which don't.

ojwb committed:
Use setlimit to avoid most $(len > N) checks

In UTF-8 mode `len` recomputes the length of the string each time.

ojwb committed:
Handle all adjective cases in one `among`

`among` currently has complexity O(log(n)) in the number of strings
(and I'm working on an O(1) implementation) so in general one big among
is more efficient than several small ones.

ojwb committed:
Handle all noun cases in one `among`

This reveals that the -MAN, -TAN and -SHAN rules couldn't actually
ever trigger before.  I've commented them out for now to preserve
the currently expected output.

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,392 @@
+// FIXME: These were dominated by -AN below before refactoring.
+//        '{mim}{alef}{nun}'  (delete) // -MAN
+//        '{te}{alef}{nun}'   (delete) // -TAN
+//        '{shin}{alef}{nun}' (delete) // -SHAN
```
Before I refactored we had:

```
    N_Hash or Noun_Other or Noun_Plural or Noun_Possessive
```

`Noun_Plural` had a rule for `'{alef}{nun}'` which would also match any word with one of the 3 commented out endings above.   (And note that the original condition on `'{alef}{nun}'` was `$(len > 4)` and the original conditions on the rules above were all `$(len > 5)`.)

I suspect we want to uncomment these, but that affects the stemming of quite a lot of words so needs reviewing.

ojwb committed:
Eliminate protect_lex_an flag

Change Protect_Lexical_AN to signal f if the word ends in a lexical
-an, and use that to exit the suffix peeling loop in that situation
since we won't be able to remove any more suffixes.

ojwb committed:
Test for {dal} or {te} before {heh} in the among

ojwb committed:
Replace all `$(len > N)` for suffixes

Either use an R1 test or `not atlimit`.

ojwb committed:
Eliminate remaining $(len > N)

ojwb committed:
Refactor flags so we only need to test one flag

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,385 @@
+    )
+  )
+
+  // --- Verb: person/number endings
```
I can see "person" endings handled below, but where are the "number" endings?

ojwb committed:
Merge verb handling into two amongs

ojwb committed:
Merge Noun and Adjective handling

saeiddrv committed:
Document why -MAN/-TAN/-SHAN possessive rules are disabled

These 3-char suffixes are ambiguous with plurals of nouns whose root
ends in the same letter (mim/te/shin). For example:
  - mardom+an (people plural) would wrongly stem to mard via -MAN
  - naqqash+an (painters) would wrongly stem to naqqaa via -SHAN
  - derakht+an (trees) would wrongly stem to derakh via -TAN

Without a lexicon the two cases cannot be distinguished. The -AN rule
handles both correctly at the cost of missing possessive stripping.

saeiddrv committed:
Move -AT suffix to plural section and clarify its primary role

-AT ({alef}{te}) is primarily an Arabic sound plural suffix
(e.g. ettela'aat, taghyiraat, entekhabaat) rather than a possessive.
Move it to the plural section and update the comment accordingly.
This is a comment/organisation change only — no stemming output changes.

saeiddrv committed:
Remove Persian/Arabic characters from code comments

Replace all Persian characters in inline comments with Latin
transliterations to avoid display issues in editors with Arabic
shaping enabled (e.g. neovim with arabicshape option):
- biroon, bishtar, milaadi, miyaan (instead of Persian examples)
- onvaan, divaan, javaan, ravaan (for -van pattern)
- NEMI/MI/NA/BI + ZWNJ (instead of invisible ZWNJ characters)
- past part. -deh/-teh (instead of Persian script)

saeiddrv committed:
Simplify Verb_Person_Endings comment to 'person endings'

'Number' here referred to grammatical number (singular/plural) which
is fused with person in Persian verb suffixes, not a separate marker.
The term caused confusion so simplified to just 'person endings'.

saeiddrv committed:
Remove na- and bi- negating prefix stripping

These are negating derivational prefixes that create words with
opposite meanings (na+dorost = incorrect, bi+khatr = safe).
Stripping them conflates opposites in search results which harms
precision. Also noted that 'na' never uses ZWNJ in practice since
alef does not join on the left in Arabic script.

ojwb committed:
Enable rule to remove -TARIN

saeiddrv committed:
Clarify verb person endings comment

Update to 'person and singular/plural endings' to make clear that
these suffixes encode both person and number (singular/plural).
Persian only has singular and plural, no dual.

ojwb commented:
The Ada failure looks like a bug in Snowball (e.g. our Ada runtime code and/or the Ada code we generate).  So nothing wrong with the code in this PR, but unfortunately a blocker for merging.  It's a pre-condition failing, so potentially we could relax or drop that pre-condition, but the pre-condition seems valid so really should be allowed to remain - I think there's a deeper problem.

I've managed to reduce to this testcase (in case you're curious, and so I can't lose this reproducer):

```
stringescapes { }

externals ( stem )

define stem as (
  setlimit tomark size - 4 for [
  backwards (
    ['AB'] (delete)
  )
)
```

with input:

```
ABCDAEAB
AHFIAJ
ABI
```

This reduced testcase looks like valid Snowball to me, but also seems to fail for C (unlike the real `persian.sbl`).

It seems to actually fail on the second entry but the other two need to be there (odd that the entry after matters).  The need for multiple inputs to trigger suggests something isn't getting reset/set when we start a new word, but if so I've not spotted where so far.

I'll debug further and sort this out.

ojwb commented:
> I'll debug further and sort this out.

Turns out the code generated to handle `setlimit tomark X for Y` in a forwards context was buggy for all languages (the limit was incorrectly restored after doing Y).  I've fixed that, but the `make check_ada_persian` still fails so it seems this bug wasn't the cause (or not the full cause).  Good to have found and fixed it though.

(`setlimit tomark X for Y` is used fairly extensively in existing stemmers, but always in a backwards context.)

I'll keep investigating.

ojwb commented:
Those Ada codegen bugs are now fixed.

I've pushed a rebase of the branch onto the latest master (with a couple of additional clean-up commits) to my fork to test these fixes actually fix the failure for Persian

https://github.com/ojwb/snowball/commits/add-persian-stemmer/

and CI now passes for all target languages:

https://github.com/ojwb/snowball/actions/runs/24297535548

saeiddrv commented:
Thank you so much for all your help and the great effort you've put into this PR — I really appreciate it!

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,393 @@
+  define N_Hash as (
+    [substring] among (
+      '{alef}{khe}{be}{alef}{re}'      (<- '{khe}{be}{re}')
+      '{khe}{vav}{alef}{sin}'          (<- '{khe}{alef}{sin}{heh}')
```
I tested with a (currently terribly hacky) coverage feature I've been working on and it seems that the `'{khe}{vav}{alef}{sin}'` case here is not exercised by anything in the testdata.  A simple demonstration is that the tests still pass with this change:
```
--- a/algorithms/persian.sbl
+++ b/algorithms/persian.sbl
@@ -244,7 +244,7 @@ backwardmode (
   define Irregular_Noun as (
     [substring] among (
       '{alef}{khe}{be}{alef}{re}'      (<- '{khe}{be}{re}')
-      '{khe}{vav}{alef}{sin}'          (<- '{khe}{alef}{sin}{heh}')
+      '{khe}{vav}{alef}{sin}'          (<- 'THISISNEVERUSED!')
       '{alef}{sin}{alef}{te}{ye}{dal}' (<- '{alef}{sin}{te}{alef}{dal}')
     )
   )
```

`Stem_Verb` below deletes suffix `{alef}{sin}` but that's tested *after* this so I don't think that is causing us not to match here.  Is there a typo, or is this just an uncommon case?

If it's correct it'd be good to add at least one real word to `voc.txt` which exercises it.

saeiddrv replied:
You're right — no word in the vocabulary ends with خواس, and no pipeline path produces this ending through suffix stripping either. Removed in 1213698.

ojwb commented:
I think we're mostly there, but one other thing:

Currently the protection against overstemming short words just counts characters, but most of the stemmers instead use a "measure" which approximates counting syllables (see https://snowballstem.org/texts/r1r2.html for some details, and some English examples) - this generally found seems to work better than a simple character count.

This is usually done by counting alternating spans of vowels and consonants, but it seems written Persian typically omits short vowels which probably makes this more complex as the counting may need to discriminate a consonant cluster from consonants with an omitted short vowel in between.  (Also based on my limited reading it seems some letters can be a vowel or consonant depending on context - perhaps somewhat like English _y_).

The length checks in the stemmer code are now expressed in terms of `p1` so if a suitable syllable count is possible we can just set `p1` based on it and pretty much everything will then use it.

saeiddrv commented:
Thanks for the detailed suggestion — I looked into implementing a syllable-based R1 measure using the visible long vowel letters (آ، ا، و، ی) with a standard `gopast vowel / gopast non-vowel` pattern and falling back to `hop 3` when no visible vowel transition is found.

After testing it against common Persian words, I found the improvement is marginal at best. The core issue is that written Persian routinely omits short vowels (a, e, o), so a large portion of common stems contain no visible vowel letters at all — words like دست (dast, "hand"), مرد (mard, "man"), سخت (sakht, "hard"), بچه (bachche, "child"). For all of these, the visible-vowel R1 finds no transition and falls back to the character count anyway.

On top of that, و and ی are ambiguous — they can be consonants (/v/ and /y/) depending on position. Treating them unconditionally as vowels would set R1 too early for words like ورزش (varzesh, "sport") or یکی (yeki, "one"), giving *less* protection than the current `hop 3`.

So the syllable-counting approach that works well for European languages doesn't really translate well to Persian orthography due to the hidden short vowels. I think `hop 3` is a reasonable default for now.

saeiddrv committed:
Remove untriggered irregular noun rule for خواس

No word in the test vocabulary ends with خواس (khe-vav-alef-sin),
and no pipeline path produces this ending through suffix stripping.
The rule was never exercised.

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,392 @@
+    [substring] among (
+      '{ar_kaf}'                                                 (<- '{kaf}')
+      '{ar_ye}' '{ar_ye_with_hamza_above}'                       (<- '{ye}')
+      '{ar_heh}' '{ar_he_marbuta}'                               (<- '{heh}')
```
This is mostly cosmetic, but having the code and description match is very helpful for people to understand the algorithm, and especially for future maintenance.

This rule in the description is:

> Arabic te marbuta (U+0629) / heh with hamza (U+06C1) → Persian heh (U+0647),

One difference is that the order is switched and it would be better if they matched - is one way round more natural?

But also the naming of `U+0629` doesn't match (`ar_he_marbuta` vs "Arabic te marbuta").  ``UnicodeData.txt` lists `U+0629` as:

```
0629;ARABIC LETTER TEH MARBUTA;Lo;0;AL;;;;;N;ARABIC LETTER TAA MARBUTAH;;;;
```

so maybe `ar_he_marbuta` is a typo?

Even if both names are valid, it would be more helpful to use matching names in code and description.

saeiddrv replied:
You're right, renamed `ar_he_marbuta` to `ar_teh_marbuta` to match the Unicode name and reordered to match the description in 52defb3.

ojwb commented on algorithms/persian.sbl:
```diff
@@ -0,0 +1,392 @@
+        '{be}{alef}{nun}'   (delete) // -BAN
+        '{gaf}{ye}'         (delete) // -GI (abstract noun)
+        '{ye}{te}'          (delete) // -YAT
+        '{ye}{ye}'          (delete) // -YY (double Y; also adj. relative -Y)
```
The description uses _-yi_ instead of `-YY` - we should pick a consistent name to use.

saeiddrv replied:
Good catch. Updated the website description to use `-yy` to match the code in 9c48c56.

ojwb commented:
> Thanks for the detailed suggestion — I looked into implementing a syllable-based R1 measure using the visible long vowel letters (آ، ا، و، ی) with a standard `gopast vowel / gopast non-vowel` pattern and falling back to `hop 3` when no visible vowel transition is found. [...]

Thanks for looking into this.  I've summarised the relevant parts of your comment into a new section in the snowball-website PR.

> On top of that, و and ی are ambiguous — they can be consonants (/v/ and /y/) depending on position. Treating them unconditionally as vowels would set R1 too early for words like ورزش (varzesh, "sport") or یکی (yeki, "one"), giving _less_ protection than the current `hop 3`.

FWIW, the letter _y_ in English can also be a consonant of vowel - Snowball's rule there is that it's not a vowel at the start of a word or after a vowel (so e.g. it's not a vowel in _yoke_ or _buy_, but is in _bye_, _cry_ and _tyre_).  Treating as a non-vowel at the start of the word seems to work for those two examples, but perhaps it's not as easy as _y_ in English, and this doesn't help address omitted short vowels which seems a hard problem.

saeiddrv committed:
Fix naming mismatches between code and description

- Rename ar_he_marbuta to ar_teh_marbuta to match Unicode name
  (U+0629 ARABIC LETTER TEH MARBUTA)
- Reorder among to match website description order

saeiddrv committed:
Keep nemi- prefix instead of stripping it, remove repeat

- nemi- (negation) is now detected to set the flag but not removed,
  consistent with how na- and bi- are already handled. Stripping
  negation conflates opposite meanings.
- Remove repeat loop since double mi- prefix is just junk (only
  one occurrence in 400K+ words).

ojwb commented:
@saeiddrv Thanks for all your work on this.

One other question - I'm in the process of trying to sort out a more consistent approach to handling apostrophes and hyphens across the stemmers (#187).  Does Persian use apostrophes or hyphens?

saeiddrv commented:
@ojwb  Thank you so much for your guidance and support throughout this, it truly wouldn't have been possible without your help.

No, Persian doesn't use apostrophes or hyphens, so it shouldn't be affected by #187.
